hypothesis-python/src/hypothesis/internal/floats.py

```diff
@@ -18,8 +18,53 @@
 from __future__ import division, print_function, absolute_import
 
 import math
+import struct
 
-from hypothesis.internal.compat import struct_pack, struct_unpack
+from hypothesis.internal.compat import CAN_PACK_HALF_FLOAT, quiet_raise, \
+    struct_pack, struct_unpack
+
+try:
+    import numpy
+except ImportError:
+    numpy = None
+
+
+STRUCT_FORMATS = {
+    16: (b'!H', b'!e'),  # Note: 'e' is new in Python 3.6, so we have helpers
+    32: (b'!I', b'!f'),
+    64: (b'!Q', b'!d'),
+}
+
+
+def reinterpret_bits(x, from_, to):
+    try:
+        buf = struct_pack(from_, x)
+    except struct.error:  # pragma: no cover
+        if from_ == b'!e' and numpy is not None and not CAN_PACK_HALF_FLOAT:
```

Comment:
Against that, this case is strictly optimal for everything else but float16 on Python <=3.5.  I'd prefer to be as fast as possible for builtin floats, float64, and float32 - float16 is very very rare for computation anyway and even then it's optimised on Python 3.6+ with this code!

Reply:
You could solve that with a top level redefinition of the function at zero performance hit.

```python
def reinterpret_bits(x, from_, to):
     ... 

if numpy is None and not CAN_PACK_HALF_FLOAT:
    real_reinterpret_bits = reinterpret_bits

    def reinterpret_bits(x, from_, to):
        if x == b'!e':
            ...
       else:
           return real_reinterpret_bits(x, from_, to)
```

That way you only pay the comparison performance hit on the platform which would otherwise be getting a huge performance hit in some cases.

If you're still against that, then my recommendation would be to not support `float16` without Python >= 3.6 or numpy at all! I think unexplained silent performance degradations are very unlikely to be preferable to an error message telling the user to install numpy.